registry/storage/shares.go

```diff
@@ -174,8 +183,72 @@ func (s *sharesStorage) loadPubkeyToIndexMappings() (map[spectypes.ValidatorPK]p
 	return m, err
 }
 
+// enrichSharesWithFeeRecipients batch-enriches shares with fee recipients for better performance
+func (s *sharesStorage) enrichSharesWithFeeRecipients(shares []*types.SSVShare) error {
+	// Collect unique owners while loading shares
+	uniqueOwners := make(map[common.Address]struct{})
+	var owners []common.Address
+
+	for _, share := range shares {
+		if _, exists := uniqueOwners[share.OwnerAddress]; !exists {
+			uniqueOwners[share.OwnerAddress] = struct{}{}
+			owners = append(owners, share.OwnerAddress)
+		}
+	}
+
+	recipients, err := s.recipientReader.GetRecipientDataMany(nil, owners)
+	if err != nil {
+		return fmt.Errorf("failed to get recipient data: %w", err)
+	}
+
+	// Enrich each share with its fee recipient
+	for _, share := range shares {
+		if feeRecipient, found := recipients[share.OwnerAddress]; found {
+			share.FeeRecipientAddress = feeRecipient
+		} else {
+			// Default to owner address if no custom recipient
+			copy(share.FeeRecipientAddress[:], share.OwnerAddress.Bytes())
+		}
+	}
+
+	return nil
+}
+
+// removeFromOwnerIndex removes a share from the owner's index
+func (s *sharesStorage) removeFromOwnerIndex(owner common.Address, share *types.SSVShare) {
+	shares := s.sharesByOwner[owner]
+	for i, sh := range shares {
+		if sh == share {
```

Comment:
The logic with pointer comparison and removal seems correct, but not trivial, so IMO we need some tests for this logic

Reply:
outdated